File: src/main/java/ru/hh/typofixer/morphers/WrongKeyTable.java
package ru.hh.typofixer.morphers;

import com.google.common.base.Function;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.Iterables;
import ru.hh.typofixer.Morph;
import ru.hh.typofixer.Variant;
import ru.hh.typofixer.Word;
import java.util.List;

public class WrongKeyTable extends Morph {
  private static final String ENGLISH = "qwertyuiop[]asdfghjkl;'zxcvbnm,./QWERTYUIOP{}ASDFGHJKL:\"ZXCVBNM<>?";
  private static final String RUSSIAN = "йцукенгшщзхъфывапролджэячсмитьбю.ЙЦУКЕНГШЩЗХЪФЫВАПРОЛДЖЭЯЧСМИТЬБЮ,";

  private static final List<String> TABLES = ImmutableList.of(
     ENGLISH,
     RUSSIAN
  );

  @Override
  public Variant apply(final Word input) {
    return new Variant(Iterables.transform(TABLES, new Function<String, Word>() {
      @Override
      public Word apply(String toTable) {
        String value = input.getValue();
        char[] result = new char[value.length()];
        for (int i = 0; i < value.length(); i++) {
          boolean found = false;
          for (String fromTable : TABLES) {
            int idx = fromTable.indexOf(value.charAt(i));
            if (idx != -1) {
              result[i] = toTable.charAt(idx);
              found = true;
              break;
            }
          }
          if (!found) {
            result[i] = value.charAt(i);
          }
        }
        return new Word(new String(result));
      }
    }));
  }
}

File: src/main/scala/ru/hh/typofixer/morphers/WrongKeyTable.scala
package ru.hh.typofixer.morphers

import ru.hh.typofixer.{Morph, Variant, Word}

class WrongKeyTable extends Morph {
  val ENGLISH = "qwertyuiop[]asdfghjkl;'zxcvbnm,./QWERTYUIOP{}ASDFGHJKL:\"ZXCVBNM<>?"

  val RUSSIAN = "йцукенгшщзхъфывапролджэячсмитьбю.ЙЦУКЕНГШЩЗХЪФЫВАПРОЛДЖЭЯЧСМИТЬБЮ,"

  val TABLES = List(ENGLISH, RUSSIAN)
  val INDEXES = TABLES.map(table => table -> table.map(ch => ch -> table.indexOf(ch)).toMap).toMap

  override def apply(token: Word) = {
    val value = token.string
    new Variant(TABLES.map(toTable =>
      new Word(value.map(ch =>
        TABLES.map(fromTable => {
          val idx = INDEXES(fromTable).getOrElse(ch, -1)
          if (idx > -1)
            toTable.charAt(idx)
          else
            null
        }).filter(_ != null).headOption.getOrElse(ch)
      ).mkString)
    ))
  }
}
